### PR TITLE
fix(ports): order

### PR DESCRIPTION
**Pull request by acateland**
**Please check if the PR fulfills these requirements**

- [X] The commit(s) message(s) follows our
  [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md#commit-message-format)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What is the current behavior?** (You can also link to an open issue here)
Port order is kept in an ordered map, but for flow loading if the source doesn't garanties the ordering, port order could get mixed.

**What is the new behavior?**
Instead a index is now generated for each port inside ```port.graphicalAttributes.properties.index```
The index is calculated by port attached to a node and by direction of the port 'SINK' OR 'EMITTER'.

on port injection the index is calculated following the lenght of the Port map attached to the same node with same direction.

on port deletion the index is recalculated for the whole Port map attached to the same node with same direction.

**Does this PR introduce a breaking change?**

- [x] Yes
- [ ] No

on redux action all occurence of portId are replaced by id 

**Other information**:

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -88,6 +90,7 @@
     "d3-selection": "1.0.3",
     "d3-shape": "1.0.3",
     "d3-zoom": "1.0.2",
+    "flow-typed": "^2.0.0",
```
Comment (jmfrancois) [inline package.json]: should be devdeps

File: src/actions/port.actions.js [hunk 2]
```diff
@@ -1,29 +1,26 @@
...
  * @param {Object} attributes
  */
-export const addPort = (nodeId, portId, { data = {}, graphicalAttributes = {} } = {}) => ({
-	type: FLOWDESIGNER_PORT_ADD,
-	nodeId,
-	portId,
-	data,
-	graphicalAttributes,
-});
+export function addPort(nodeId: string, id: string, { data, graphicalAttributes }: {data: PortData, graphicalAttributes: PortGraphicalAttributes}): PortActionAdd {
```
Comment (jmfrancois) [inline src/actions/port.actions.js]: style: lines < 100

File: src/actions/port.actions.js [hunk 3]
```diff
@@ -1,29 +1,26 @@
...
+		nodeId,
+		id,
+		data,
+		graphicalAttributes,
+	};
+}
 
-export const addPorts = (nodeId, ports) => ({
-	type: FLOWDESIGNER_PORT_ADDS,
+export const addPorts = (nodeId: string, ports: Array<Port>): PortAction => ({
```
Comment (jmfrancois) [inline src/actions/port.actions.js]: please write a function

File: src/actions/port.actions.js [hunk 4]
```diff
@@ -44,8 +41,8 @@ export const setPortGraphicalAttribute = (portId, graphicalAttributes) => ({
  * @param {string} portId
  * @param {string} graphicalAttributesKey - the key of the attribute to be removed
  */
-export const removePortGraphicalAttribute = (portId, graphicalAttributesKey) => ({
-	type: FLOWDESIGNER_PORT_REMOVE_GRAPHICAL_ATTRIBUTES,
+export const removePortGraphicalAttribute = (portId: string, graphicalAttributesKey: string): PortAction => ({
+	type: 'FLOWDESIGNER_PORT_REMOVE_GRAPHICAL_ATTRIBUTES',
```
Comment (jmfrancois) [inline src/actions/port.actions.js]: You don't use constant anymore.
could you ensure the constant is also checked by flow type ?

File: src/actions/port.actions.test.js [hunk 5]
```diff
@@ -1,3 +1,5 @@
+
+/* @flow */
 /* eslint-disable import/no-extraneous-dependencies */
 import configureMockStore from 'redux-mock-store';
 import thunk from 'redux-thunk';
```
Comment (jmfrancois) [inline src/actions/port.actions.test.js]: do you really need thunk here ?

File: src/reducers/port.reducer.js [hunk 6]
```diff
@@ -17,58 +22,102 @@ import {
...
+	const index: number = port.graphicalAttributes.properties.index || calculateNewPortIndex(state.get('ports'), port);
 	const newState = state.setIn(['ports', port.id], new PortRecord({
 		id: port.id,
 		nodeId: port.nodeId,
 		data: new PortData(port.data)
 			.set('properties', fromJS(port.data && port.data.properties) || new Map()),
 		graphicalAttributes: new PortGraphicalAttributes(port.graphicalAttributes)
 			.set('position', new PositionRecord(port.graphicalAttributes.position))
-			.set('properties', fromJS(port.graphicalAttributes && port.graphicalAttributes.properties) || new Map()),
+			.set('properties', fromJS(port.graphicalAttributes && { index, ...port.graphicalAttributes.properties }) || new Map()),
```
Comment (jmfrancois) [inline src/reducers/port.reducer.js]: style: len > 100 (122)

File: src/reducers/port.reducer.js [hunk 7]
```diff
@@ -17,58 +22,102 @@ import {
...
+		}
+		return 0;
+	}).map((port) => {
+		i += 1;
+		return port.setIndex(i - 1);
+	});
+}
+
+function setPort(state: State, port: Port) {
+	const index: number = port.graphicalAttributes.properties.index || calculateNewPortIndex(state.get('ports'), port);
```
Comment (jmfrancois) [inline src/reducers/port.reducer.js]: style: len > 100 (116)

File: src/reducers/port.reducer.js [hunk 8]
```diff
@@ -101,21 +150,27 @@ export default function portReducer(state = defaultState, action) {
...
 					(cumulativeState, link) => linkReducer(cumulativeState, removeLink(link.id)),
 					state,
 				),
 			)
 			.deleteIn(['ports', action.portId])
-			.deleteIn(['out', state.getIn(['ports', action.portId]).nodeId, action.portId])
-			.deleteIn(['in', state.getIn(['ports', action.portId]).nodeId, action.portId]);
+			.deleteIn(['out', state.getIn(['ports', action.portId, 'nodeId']), action.portId])
+			.deleteIn(['in', state.getIn(['ports', action.portId, 'nodeId']), action.portId]);
+			return newState.mergeDeep({ ports: indexPortMap(filterPortsByDirection(filterPortsByNode(newState.get('ports'), port.nodeId), port.getPortDirection())) });
```
Comment (jmfrancois) [inline src/reducers/port.reducer.js]: style: len > 100 (158) ⚠️

File: src/flow-typed/index.js [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+	getNodeType: () => string,
+};
+
+export type LinkRecordType = Record<*>
+& {
+	getLinkType: () => string,
+};
+
+/** $STATE */
+
```
Comment (jmfrancois) [inline src/flow-typed/index.js]: 73:1  error  More than 2 blank lines not allowed  no-multiple-empty-lines

File: package.json (hunk 1 above)
Comment (jmfrancois) [inline package.json]: should be devdeps

File: src/actions/port.actions.js (hunk 2 above)
Comment (jmfrancois) [inline src/actions/port.actions.js]: style: lines < 100

File: src/components/node/AbstractNode.component.js [hunk 10]
```diff
@@ -32,15 +31,31 @@ function calculatePortPosition(ports, nodePosition, nodeSize) {
 		.range(range);
 	let emitterNumber = 0;
 	let sinkNumber = 0;
-	emitterPorts.forEach((port) => {
+	emitterPorts.sort((a, b) => {
+		if (a.getIndex() < b.getIndex()) {
+			return -1;
+		}
+		if (a.getIndex() > b.getIndex())	{
```
Comment (talend-glorieux) [inline src/components/node/AbstractNode.component.js]: So much space ;)

File: src/components/node/AbstractNode.component.js [hunk 11]
```diff
@@ -32,15 +31,31 @@ function calculatePortPosition(ports, nodePosition, nodeSize) {
...
 			y: scaleYEmitter(emitterNumber),
 		});
 		portsWithPosition = portsWithPosition.set(port.id, port.setIn(['graphicalAttributes', 'position'], position));
 	});
-	sinkPorts.forEach((port) => {
+	sinkPorts.sort((a, b) => {
+		if (a.getIndex() < b.getIndex()) {
+			return -1;
+		}
+		if (a.getIndex() > b.getIndex())	{
```
Comment (talend-glorieux) [inline src/components/node/AbstractNode.component.js]: So much space ;)

File: src/flow-typed/index.js [hunk 12]
```diff
@@ -0,0 +1,120 @@
+/* @flow */
+import type { Record, Map } from 'immutable';
+
+/** $BASIC */
```
Comment (talend-glorieux) [inline src/flow-typed/index.js]: Is the `$` something from flow?

File: src/flow-typed/index.js [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+
+export type LinkRecordType = Record<*>
+& {
+	getLinkType: () => string,
+};
+
+/** $STATE */
+
+
+
```
Comment (talend-glorieux) [inline src/flow-typed/index.js]: Extra newlines.

File: src/flow-typed/index.js [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+type getStateNodes = (selector: ['nodes', Id]) => NodeRecordType;
+type getStatePorts = (selector: ['ports', Id]) => PortRecordType;
+type getStateLinks = (selector: ['links', Id]) => Record<*>;
+
+export type State = Map<string, Map<Id, any>> & {
+	+getIn: getStateNodes | getStatePorts | getStateLinks
+};
+
+
```
Comment (talend-glorieux) [inline src/flow-typed/index.js]: Extra newline

File: src/reducers/port.reducer.js [hunk 15]
```diff
@@ -17,58 +22,102 @@ import {
...
+	).size;
+}
+
+function indexPortMap(ports: PortRecordMap): PortRecordMap {
+	let i = 0;
+	return ports.sort((a, b) => {
+		if (a.getIndex() < b.getIndex()) {
+			return -1;
+		}
+		if (a.getIndex() > b.getIndex())	{
```
Comment (talend-glorieux) [inline src/reducers/port.reducer.js]: Soo much space ;)

File: src/actions/port.actions.test.js (hunk 5 above)
Comment (acateland) [inline src/actions/port.actions.test.js]: no good catch !

File: src/flow-typed/index.js (hunk 12 above)
Comment (acateland) [inline src/flow-typed/index.js]: no its just an helper for searching trought comment sections

Comment (talend-glorieux) [inline src/flow-typed/index.js]: Not sure it helps that much.

Comment (acateland) [inline src/flow-typed/index.js]: it doest when file start to get quite long

**Pull request merged**